examples/android/clientcache/app/src/main/java/io/grpc/clientcacheexample/SafeMethodCachingInterceptor.java

```diff
@@ -0,0 +1,297 @@
+
+    void clear();
+  }
+
+  /**
+   * Obtain a new cache with a least-recently used eviction policy and the specified size limit. The
+   * backing caching implementation is provided by {@link LruCache}. It is safe for a single cache
+   * to be shared across multiple {@link SafeMethodCachingInterceptor}s without synchronization.
+   */
+  public static Cache newLruCache(final int cacheSizeInBytes) {
+    return new Cache() {
+      private final LruCache<Key, Value> lruCache =
+          new LruCache<Key, Value>(cacheSizeInBytes) {
+            protected int sizeOf(Key key, Value value) {
+              return value.response.getSerializedSize();
+            }
+          };
+
+      @Override
+      public void put(Key key, Value value) {
+        lruCache.put(key, value);
+      }
+
+      @Override
+      public Value get(Key key) {
+        return lruCache.get(key);
+      }
+
+      @Override
+      public void remove(Key key) {
+        lruCache.remove(key);
+      }
+
+      @Override
+      public void clear() {
+        lruCache.evictAll();
+      }
+    };
+  }
+
+  public static SafeMethodCachingInterceptor newSafeMethodCachingInterceptor(Cache cache) {
+    return newSafeMethodCachingInterceptor(cache, DEFAULT_MAX_AGE_SECONDS);
+  }
+
+  public static SafeMethodCachingInterceptor newSafeMethodCachingInterceptor(
+      Cache cache, int defaultMaxAge) {
+    return new SafeMethodCachingInterceptor(cache, defaultMaxAge);
+  }
+
+  private static int DEFAULT_MAX_AGE_SECONDS = 3600;
+
+  private static final Metadata.Key<String> CACHE_CONTROL_KEY =
+      Metadata.Key.of("cache-control", Metadata.ASCII_STRING_MARSHALLER);
+
+  private static final Splitter CACHE_CONTROL_SPLITTER =
+      Splitter.on(',').trimResults().omitEmptyStrings();
+
+  private final Cache internalCache;
+  private final int defaultMaxAge;
+
+  private SafeMethodCachingInterceptor(Cache cache, int defaultMaxAge) {
+    this.internalCache = cache;
+    this.defaultMaxAge = defaultMaxAge;
+  }
+
+  @Override
+  public <ReqT, RespT> ClientCall<ReqT, RespT> interceptCall(
+      final MethodDescriptor<ReqT, RespT> method, final CallOptions callOptions, Channel next) {
+    // Currently only unary methods can be marked safe, but check anyways.
+    if (!method.isSafe() || method.getType() != MethodDescriptor.MethodType.UNARY) {
+      return next.newCall(method, callOptions);
+    }
+
+    final String fullMethodName = method.getFullMethodName();
+
+    return new ForwardingClientCall.SimpleForwardingClientCall<ReqT, RespT>(
+        next.newCall(method, callOptions)) {
+      private Listener<RespT> interceptedListener;
+      private Key requestKey;
+      private boolean cacheResponse = true;
+      private boolean cacheOptionsCausedCancellation;
+
+      @Override
+      public void start(Listener<RespT> responseListener, Metadata headers) {
+        interceptedListener =
+            new ForwardingClientCallListener.SimpleForwardingClientCallListener<RespT>(
```

Comment:
optional: there is a ton of indenting in this function.  maybe extract this to a helper class?

Reply:
There's a lot of shared state; refactored to a helper class but it looks cleaner as-is so reverted.